File: scripts/crawl-geojson.js
/**
 * 从阿里云抓取各区域的 geojson
 */
import { readdir, readFile, rm, writeFile } from 'node:fs/promises'
import { createRequire } from 'node:module'
import { dirname, extname, resolve } from 'node:path'
import { fileURLToPath } from 'node:url'

import axios from 'axios'

function getGeoPath() {
  const __filename = fileURLToPath(import.meta.url)
  const __dirname = dirname(__filename)
  const geoPath = resolve(__dirname, '../public/json/geo')
  return geoPath
}

function getHasGeoAreas() {
  const _require = createRequire(import.meta.url)
  const list = _require('../public/json/all-area.json')
  const filteredList = list.filter((item) => item.level !== 'district')
  return filteredList
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms))
}

async function prepareGeoFiles() {
  const existsFiles = []
  try {
    const geoPath = getGeoPath()
    const files = await readdir(geoPath)
    for (let f of files) {
      const filepath = resolve(geoPath, f)
      const file = await readFile(filepath)
      const fileExtname = extname(filepath)
      // 空文件就删除
      if (file.byteLength <= 0 || !fileExtname.includes('json')) {
        await rm(filepath)
      } else {
        existsFiles.push(f)
      }
    }
  } catch (err) {
    console.error(err)
  }
  const hasGeoAreas = await getHasGeoAreas()
  const requestList = hasGeoAreas.filter(
    (item) => !existsFiles.find((f) => f.includes(item.adcode + ''))
  )
  return requestList
}

async function catchGeoJson() {
  try {
    const requestList = await prepareGeoFiles()
    if (!requestList.length) {
      console.log('Not need to catched geojson data.')
      return
    }
    for (const reqItem of requestList) {
      try {
        const res = await axios.get(
          `https://geo.datav.aliyun.com/areas_v3/bound/${reqItem.adcode}_full.json`
        )
        if (res.status === 200) {
          const data = JSON.stringify(res.data)
          const geoPath = resolve(getGeoPath(), reqItem.adcode + '.json')
          await writeFile(geoPath, data)
          console.log(`Success write data to ${geoPath}`)
        }
      } catch (err) {
        console.error(JSON.stringify(err))
      }
      await sleep(1500)
    }
  } catch (err) {
    console.error(JSON.stringify(err))
  }
}

catchGeoJson()

File: scripts/crawl-geojson.go
package main

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"os"
	"time"
)

type AreaLevel string

const (
	AreaLevelProvince AreaLevel = "province"
	AreaLevelCity     AreaLevel = "city"
	AreaLevelDistrict AreaLevel = "district"
)

type Area struct {
	Name   string    `json:"name"`
	Adcode int       `json:"adcode"`
	Parent int       `json:"parent"`
	Level  AreaLevel `json:"level"`
	Lng    float64   `json:"lng"`
	Lat    float64   `json:"lat"`
}

const (
	geojsonPath = "public/json/geo"
)

func main() {
	makeGeojsonDir()
	adcodes := getAreaAdcodes()
	crawl(adcodes)
}

// 错误处理
func check(e error) {
	if e != nil {
		fmt.Println(e)
	}
}

// 创建 geojson 文件夹
func makeGeojsonDir() {
	if _, err := os.Stat(geojsonPath); os.IsNotExist(err) {
		err := os.Mkdir(geojsonPath, 0755)
		check(err)
	}
}

// 写入 geojson 到文件
func writeGeojson(adcode int, data []byte) {
	filename := fmt.Sprintf("%d.json", adcode)
	filepath := fmt.Sprintf("%s/%s", geojsonPath, filename)
	f, err := os.Create(filepath)
	check(err)

	defer f.Close()

	fb, err := f.Write(data)
	check(err)
	fmt.Printf("%s wrote %d bytes\n", filename, fb)
}

// 爬取 geojson
func getAreaGeojson(adcode int) ([]byte, error) {
	reqUrl := fmt.Sprintf("https://geo.datav.aliyun.com/areas_v3/bound/%d_full.json", adcode)
	resp, err := http.Get(reqUrl)
	check(err)

	if resp.StatusCode != 200 {
		return nil, errors.New(fmt.Sprintf("crawl geojson failed, statusCode: %d, status: %s\n", resp.StatusCode, resp.Status))
	}

	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	check(err)

	return body, nil
}

// 获取要抓取区域的 adcode
func getAreaAdcodes() []int {
	areasBytes, err := os.ReadFile("./public/json/all-area.json")
	check(err)

	var areas []Area
	err = json.Unmarshal(areasBytes, &areas)
	check(err)

	var adcodes []int
	for _, area := range areas {
		adcodes = append(adcodes, area.Adcode)
	}

	return adcodes
}

func crawl(adcodes []int) {
	for index, adcode := range adcodes {
		json, err := getAreaGeojson(adcode)

		if index%2 != 0 {
			time.Sleep(1 * time.Second)
		}

		if err != nil {
			fmt.Printf("%d: %v", adcode, err)
			continue
		}

		writeGeojson(adcode, json)
	}
}
